### PR TITLE
Support for reporting to multiple zabbix servers.

### PR DESCRIPTION
**Pull request by erikoest**
This is a suggestion for supporting reporting to multiple zabbix servers. The code has been tested and seems to work. The configration yaml-file is slightly changed, as shown in the example file, but the original format with a scalar zabbix_host can still be read.

### CODE REVIEW

**Comment by shamil** [comment]
Hey @erikoest,

I think the implementation of sending reports to multiple Zabbix servers should be in the reports logic itself, and not in `Puppet::Util::Zabbix::Sender` utility.

Would you like to make the required changes?

**Comment by shamil** [comment]
Relates to #3

**Comment by erikoest** [comment]
I understand. I made the changes in the Sender class because it felt
simpler, i.e. less changes. I can try to put it in the reports logic
instead if you want that.

Regards,
Erik Østlyngen

**Comment by shamil** [comment]
> I can try to put it in the reports logic instead if you want that.

@erikoest , that would be awesome...

File: zabbix.yaml [hunk 1]
```diff
@@ -1,5 +1,7 @@
 ---
-:zabbix_host: zabbix.domain.tld
+:zabbix_host:
+  - zabbix1.domain.tld
+  - zabbix2.domain.tld
 :zabbix_port: 10051
```
Comment (shamil) [inline zabbix.yaml:5]: Having multiple hosts is also means we need to be able to set `port` per each host, no?
You could wrap this with separate field, like

``` yaml
:zabbix_hosts:
- addr: zabbix1.domain.tld
  port: 10051
- addr: zabbix2.domain.tld
  port: 10051
```
- The port can be optional and can inherit from `:zabbix_port` or default to `10051`
- To keep backward compatibility, you keep `:zabbix_host` and `:zabbix_port`, and in the report logic you just convert those to the newer structure

File: lib/puppet/reports/zabbix.rb [hunk 2]
```diff
@@ -11,28 +11,41 @@ def process
...
-    # send metrics to zabbix
-    Puppet.debug "sending zabbix report for host #{self.host}, at #{zabbix_sender.serv}:#{zabbix_sender.port}"
-    result = zabbix_sender.send! host_overrides.fetch(self.host, self.host)
+      # send metrics to zabbix
+      Puppet.debug "sending zabbix report for host #{self.host}, at #{zabbix_sender.serv}:#{zabbix_sender.port}"
+      result = zabbix_sender.send! host_overrides.fetch(self.host, self.host)
+
+      # validate the response. if it fails, keep on sending to all
+      # zabbix servers before reporting the error.
+      if result['response'] != 'success2'
```
Comment (shamil) [inline lib/puppet/reports/zabbix.rb:44]: what is `success2`?

**Review by shamil — CHANGES_REQUESTED** [review]

File: lib/puppet/reports/zabbix.rb [hunk 3]
```diff
@@ -11,28 +11,41 @@ def process
...
+        raise_error ||= result['info']
```
Comment (shamil) [inline lib/puppet/reports/zabbix.rb:45]: I recommend just print the error, like `Puppet.error "zabbix send failed - #{result['info']}" if result['response'] != 'success'`  and set `raise_error = true`

later after exiting the each loop you can raise exception telling that `zabbix report had failures` or something similar...

**Comment by erikoest** [comment]
Hi,

I messed up my git, so I deleted it and started anew. I will send you a
new merge request when I have a cleaner history. I can add multiple
ports also. Less important than multiple hosts, but nice to have, I
guess.

/ Erik

**Comment by shamil** [comment]
Closing in favor of #5

**Pull request closed without merging**